### ISSUE TITLE
unreachable code not deteched when `never` assigned to a variable

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#45267]
# Bug Report

### 🔎 Search Terms

unreachable code

### 🕗 Version & Regression Information
4.4.0-dev.20210801

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-dev.20210801#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlRADcQYBYAKCbDwGcN4p4BeA4ki3Y4IIAHQQcAcyIBGEvAD0i1Dnh0YOBs0aJCpJkNRsR4yTPlLFGrUA)

### 💻 Code

```ts
declare function foo(): never

const a = foo()
console.log(1) //no error

foo()

console.log(1) //error
```

### 🙁 Actual behavior

error on code after `never` is assigned to `a`

### 🙂 Expected behavior

no error

**Comment by KotlinIsland** [comment]
Also this:
```ts
function TODO(): never { throw new Error }
declare function unreachable<A>(): A extends number ? never : A


unreachable<10>()
console.log("cringe") // no error
TODO()
console.log("cringe") //  yes error
```

**Comment by DetachHead** [comment]
another example:
```ts
declare const foo: 1|2|3

switch(foo) {
    case 1:
        console.log(foo)
        break
    case 2:
        console.log(foo)
        break
    case 3:
        console.log(foo)
        break
    default:
        console.log(foo) //no error
}
```

**Comment by KotlinIsland** [comment]
> another example:
> 
> ```ts
> declare const foo: 1|2|3
> 
> switch(foo) {
>     case 1:
>         console.log(foo)
>         break
>     case 2:
>         console.log(foo)
>         break
>     case 3:
>         console.log(foo)
>         break
>     default:
>         console.log(foo) //no error
> }
> ```
> 
> https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&noUnusedLocals=true&noUnusedParameters=true&noPropertyAccessFromIndexSignature=true&noImplicitOverride=true&noFallthroughCasesInSwitch=true#code/CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAIwA+ATCQMwCwAUHWgO4CWGYAFgBS5ICU8AbzrwRiKGgRF8w0bOTokEEADoISAObc8vGbJEAjOFADWusRPhlptPaPlpFKtZp46bt+IZAmzYcQkprD0RUByVVDS0+M1kvH3cRUGwoAFcIDCCPe0cIl214AHoClCR4EBgYJBg6AF8gA

This is a different issue, not related to assignment of a never function.

**Comment by andrewbranch** [comment]
That’s working as intended; `never` is assignable to `any`.

**Comment by KotlinIsland** [comment]
> That’s working as intended; `never` is assignable to `any`.

The issue isn't that it's assignable or not, the issue is that when a `never` value is assigned, no "unreachable code" error is detected.

**Comment by andrewbranch** [comment]
It’s nice to be able to write an `assertNever(foo)` there, which gives you a compile-time guarantee that your switch statement is exhaustive and a runtime error explaining what value unexpectedly made it through. https://github.com/microsoft/TypeScript/blob/79474fdfd26897eeaeb9268adfb1f4d290dc9033/src/compiler/debug.ts#L199-L202

**Comment by KotlinIsland** [comment]
I agree that assert never is indeed based, but I would like it to be opt in, the detection of unreachable code be default seems correct in my opinion.

I would expect the following behavior:
```ts
declare function foo(): true
declare function bar(): never

const a = foo();
if (a) {
  console.log(a)
} else {
  // ok, now this is based, a 'should be unreachable' branch with a check.

  // @ts-expect-error assert unreachable 
  assertNever(a)
}
const b = bar()
console.log(b) // Unreachable code detected.(7027)
doImportantCleanup()
// oh right, this is unreachable because bar returns never, good thing the compiler found this defect
//  in my code and I didn't ship this to procuction.
```

**Comment by DetachHead** [comment]
i think it would be nice if there was an exception in the logic to detect unreachable code that allows it on function calls that take `never` as a parameter. this would allow for the `assertNever` pattern while also correctly pointing out unintended unreachable code (which i feel disabling `allowUnreachableCode` currently fails to do like 90% of the time)

**Comment by RyanCavanaugh** [comment]
There are two separable concerns here:
 * Is *syntactically unreachable* code -- code which we can be certain cannot execute -- allowed? That is the only thing that `allowUnreachableCode` speaks to
 * What should be done with *type unreachable` code -- code which the type system believes will not execute? That is the realm of `assertNever`

Keeping these concepts separate, rather than conflating them, is the simpler design.

**Comment by DetachHead** [comment]
in that case, i'd propose two separate flags for each concern. perhaps `allowUnreachableNever` or something like that

**Comment by KotlinIsland** [comment]
The flag actually checks unreachable never and syntax:
```ts
declare function foo(): never

foo();

console.log("never") // error without allowUnreachableCode, no error with allowUnreachableCode
```
```ts
throw new Error;
console.log("never") // error without allowUnreachableCode, no error with allowUnreachableCode
```
